File: stock-market-service/src/main/java/rs/edu/raf/strategy/AbstractOrderProcessStrategy.java
package rs.edu.raf.strategy;

import com.google.gson.Gson;
import jakarta.transaction.Transactional;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.UpdateBalanceDTO;
import rs.edu.raf.model.ListingOwner;
import rs.edu.raf.model.order.Order;
import rs.edu.raf.model.order.OrderAction;
import rs.edu.raf.repository.ListingOwnerRepository;
import rs.edu.raf.repository.OrderRepository;

import java.util.List;

@Component
public abstract class AbstractOrderProcessStrategy implements OrderProcessStrategy{
    private String UPDATE_BALANCE_EXCHANGE;
    private String UPDATE_BALANCE_ROUTING_KEY;
    private OrderRepository orderRepository;
    private ListingOwnerRepository listingOwnerRepository;
    private RabbitTemplate rabbitTemplate;
    private Gson gson;

    public AbstractOrderProcessStrategy(@Value("${update.balance.exchange}") String UPDATE_BALANCE_EXCHANGE,
                                        @Value("${update.balance.routing.key}") String UPDATE_BALANCE_ROUTING_KEY,
                                        OrderRepository orderRepository, ListingOwnerRepository listingOwnerRepository,
                                        RabbitTemplate rabbitTemplate, Gson gson) {
        this.UPDATE_BALANCE_EXCHANGE = UPDATE_BALANCE_EXCHANGE;
        this.UPDATE_BALANCE_ROUTING_KEY = UPDATE_BALANCE_ROUTING_KEY;
        this.orderRepository = orderRepository;
        this.listingOwnerRepository = listingOwnerRepository;
        this.rabbitTemplate = rabbitTemplate;
        this.gson = gson;
    }

    /**
     * Checks if the order is executable
     * @param order the order
     * @return true if the order is executable, otherwise false
     */
    protected abstract boolean isOrderExecutable(Order order);

    protected void processBuyOrder(Order buyOrder, Double buyPrice, Double sellPrice) {

        List<Order> sellOrders = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
                buyOrder.getTicker(), OrderAction.SELL, 0);

        for(Order sellOrder: sellOrders) {
            if(buyOrder.getOrderCreator().longValue() == sellOrder.getOrderCreator().longValue())
                continue;

            if(!isOrderExecutable(sellOrder))
                continue;

            executeTransaction(buyOrder, sellOrder, buyPrice, sellPrice);

            if(buyOrder.getQuantity() == 0) {
                return;
            }
        }
    }

    protected void processSellOrder(Order sellOrder, Double buyPrice, Double sellPrice) {
        List<Order> buyOrders = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
                sellOrder.getTicker(), OrderAction.BUY, 0);

        for(Order buyOrder: buyOrders) {
            if(sellOrder.getOrderCreator().longValue() == buyOrder.getOrderCreator().longValue())
                continue;

            if(!isOrderExecutable(buyOrder))
                continue;

            executeTransaction(buyOrder, sellOrder, buyPrice, sellPrice);

            if(sellOrder.getQuantity() == 0) {
                return;
            }
        }
    }

    @Transactional
    protected void executeTransaction(Order buyOrder, Order sellOrder, Double buyPrice, Double sellPrice) {
        ListingOwner buyListingOwner = listingOwnerRepository.findByOwnerAndTicker(buyOrder.getOrderCreator(), buyOrder.getTicker())
                .orElse(new ListingOwner(buyOrder.getOrderCreator(), buyOrder.getTicker(), 0));

        ListingOwner sellListingOwner = listingOwnerRepository.findByOwnerAndTicker(sellOrder.getOrderCreator(), sellOrder.getTicker())
                .orElseThrow();

        if(sellOrder.getQuantity() <= buyOrder.getQuantity()) {
            Integer quantity = sellOrder.getQuantity();
            buyListingOwner.setQuantity(buyListingOwner.getQuantity() + quantity);
            sellListingOwner.setQuantity(sellListingOwner.getQuantity() - quantity);
            buyOrder.setQuantity(buyOrder.getQuantity() - quantity);
            sellOrder.setQuantity(0);
            updateBankAccountBalance(buyListingOwner.getOwner(), sellListingOwner.getOwner(), buyPrice * quantity,
                    sellPrice * quantity);
        }
        else {
            Integer quantity = buyOrder.getQuantity();
            buyListingOwner.setQuantity(buyListingOwner.getQuantity() + quantity);
            sellListingOwner.setQuantity(sellListingOwner.getQuantity() - quantity);
            sellOrder.setQuantity(sellOrder.getQuantity() - quantity);
            buyOrder.setQuantity(0);
            updateBankAccountBalance(buyListingOwner.getOwner(), sellListingOwner.getOwner(), buyPrice * quantity,
                    sellPrice * quantity);
        }
        orderRepository.save(sellOrder);
        orderRepository.save(buyOrder);
        listingOwnerRepository.save(buyListingOwner);
        listingOwnerRepository.save(sellListingOwner);
    }

    protected void updateBankAccountBalance(Long buyerId, Long sellerId, Double buyPrice, Double sellPrice) {
        UpdateBalanceDTO updateBalanceDTO = new UpdateBalanceDTO(buyerId, sellerId, buyPrice, sellPrice);
        rabbitTemplate.convertAndSend(UPDATE_BALANCE_EXCHANGE, UPDATE_BALANCE_ROUTING_KEY, gson.toJson(updateBalanceDTO));
    }
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/strategy/AbstractOrderProcessStrategy.kt
package org.example.stockmarketservicekotlin.strategy

import com.google.gson.Gson
import jakarta.transaction.Transactional
import org.example.stockmarketservicekotlin.dto.UpdateBalanceDTO
import org.example.stockmarketservicekotlin.model.ListingOwner
import org.example.stockmarketservicekotlin.model.order.Order
import org.example.stockmarketservicekotlin.model.order.OrderAction
import org.example.stockmarketservicekotlin.repository.ListingOwnerRepository
import org.example.stockmarketservicekotlin.repository.OrderRepository
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
abstract class AbstractOrderProcessStrategy(
    @Value("\${update.balance.exchange}") private val UPDATE_BALANCE_EXCHANGE: String,
    @Value("\${update.balance.routing.key}") private val UPDATE_BALANCE_ROUTING_KEY: String,
    protected val orderRepository: OrderRepository,
    private val listingOwnerRepository: ListingOwnerRepository,
    private val rabbitTemplate: RabbitTemplate,
    private val gson: Gson
) : OrderProcessStrategy {

    /**
     * Checks if the order is executable
     * @param order the order
     * @return true if the order is executable, otherwise false
     */
    protected abstract fun isOrderExecutable(order: Order): Boolean

    protected fun processBuyOrder(buyOrder: Order, buyPrice: Double, sellPrice: Double) {
        val sellOrders: List<Order> = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
            buyOrder.ticker, OrderAction.SELL, 0
        )

        for (sellOrder in sellOrders) {
            if (buyOrder.orderCreator == sellOrder.orderCreator) continue

            if (!isOrderExecutable(sellOrder)) continue

            executeTransaction(buyOrder, sellOrder, buyPrice, sellPrice)

            if (buyOrder.quantity == 0) {
                return
            }
        }
    }

    protected fun processSellOrder(sellOrder: Order, buyPrice: Double, sellPrice: Double) {
        val buyOrders: List<Order> = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
            sellOrder.ticker, OrderAction.BUY, 0
        )

        for (buyOrder in buyOrders) {
            if (sellOrder.orderCreator == buyOrder.orderCreator) continue

            if (!isOrderExecutable(buyOrder)) continue

            executeTransaction(buyOrder, sellOrder, buyPrice, sellPrice)

            if (sellOrder.quantity == 0) {
                return
            }
        }
    }

    @Transactional
    protected fun executeTransaction(buyOrder: Order, sellOrder: Order, buyPrice: Double, sellPrice: Double) {
        val buyListingOwner: ListingOwner =
            listingOwnerRepository.findByOwnerAndTicker(buyOrder.orderCreator, buyOrder.ticker)
                .orElse(ListingOwner(null, buyOrder.orderCreator, buyOrder.ticker, 0))

        val sellListingOwner: ListingOwner =
            listingOwnerRepository.findByOwnerAndTicker(sellOrder.orderCreator, sellOrder.ticker)
                .orElseThrow()

        if (sellOrder.quantity <= buyOrder.quantity) {
            val quantity: Int = sellOrder.quantity
            buyListingOwner.quantity += quantity
            sellListingOwner.quantity -= quantity
            buyOrder.quantity -= quantity
            sellOrder.quantity = 0
            updateBankAccountBalance(
                buyListingOwner.owner, sellListingOwner.owner, buyPrice * quantity,
                sellPrice * quantity
            )
        } else {
            val quantity: Int = buyOrder.quantity
            buyListingOwner.quantity += quantity
            sellListingOwner.quantity -= quantity
            sellOrder.quantity -= quantity
            buyOrder.quantity = 0
            updateBankAccountBalance(
                buyListingOwner.owner, sellListingOwner.owner, buyPrice * quantity,
                sellPrice * quantity
            )
        }
        orderRepository.save(sellOrder)
        orderRepository.save(buyOrder)
        listingOwnerRepository.save(buyListingOwner)
        listingOwnerRepository.save(sellListingOwner)
    }

    protected fun updateBankAccountBalance(buyerId: Long, sellerId: Long, buyPrice: Double, sellPrice: Double) {
        val updateBalanceDTO: UpdateBalanceDTO = UpdateBalanceDTO(buyerId, sellerId, buyPrice, sellPrice)
        rabbitTemplate.convertAndSend(
            UPDATE_BALANCE_EXCHANGE,
            UPDATE_BALANCE_ROUTING_KEY,
            gson.toJson(updateBalanceDTO)
        )
    }
}